File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-7.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxPathSum(root) {
        let sum = [-Infinity];
        console.log(this.postorder(root, sum));

        return sum[0]
    }

    postorder(root, sum){
        if (!root){
            return 0;
        }

        console.log(`curr node ${root.val}`)

        // if contribution is less than 0, refuse it
        let left = Math.max(this.postorder(root.left, sum), 0);
        let right = Math.max(this.postorder(root.right, sum), 0);

        // if larger- it will increase our curr path sum (aka the local path)
        let currentPathSum  = root.val + left + right;

        // compare curr max, with local path
        sum[0] = Math.max(sum[0], currentPathSum);

        // return subtree contribution to parent
        return root.val + Math.max(left, right);
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=, NoDefault0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isValid(node, minLimit, maxLimit):
            if not node:
                return True

            if not minLimit < node.val < maxLimit:
                return False

            return isValid(node.left, minLimit, node.val) and isValid(node.right, node.val, maxLimit)

        return isValid(root, float("-inf"), float("inf"))